pkg/provisioner/helper_hostpath.go

```diff
@@ -162,6 +209,105 @@ func (p *Provisioner) createCleanupPod(ctx context.Context, pOpts *HelperPodOpti
+		pOpts.hardLimitGrace == "0k" {
+		// Hack: using convertToK() style conversion
+		// TODO: Refactor this section of the code
+		pvcStorageInK := math.Ceil(float64(pOpts.pvcStorage) / 1024)
+		pvcStorageInKString := strconv.FormatFloat(pvcStorageInK, 'f', -1, 64) + "k"
+		pOpts.softLimitGrace, pOpts.hardLimitGrace = pvcStorageInKString, pvcStorageInKString
+		return nil
+	}
+
+	if pOpts.softLimitGrace == "0k" ||
+		pOpts.hardLimitGrace == "0k" {
+		return nil
+	}
+
+	if len(pOpts.softLimitGrace) > len(pOpts.hardLimitGrace) ||
+		(len(pOpts.softLimitGrace) == len(pOpts.hardLimitGrace) &&
+			pOpts.softLimitGrace > pOpts.hardLimitGrace) {
+		return errors.Errorf("hard limit cannot be smaller than soft limit")
+	}
+
+	return nil
+}
+
+// createQuotaPod launches a helper(busybox) pod, to apply the quota.
+//
+//	The local pv expect the hostpath to be already present before mounting
+//	into pod. Validate that the local pv host path is not created under root.
+func (p *Provisioner) createQuotaPod(ctx context.Context, pOpts *HelperPodOptions) error {
+	var config podConfig
+	config.pOpts, config.podName = pOpts, "quota"
+	// err := pOpts.validate()
+	if err := pOpts.validate(); err != nil {
+		return err
+	}
+
+	// Initialize HostPath builder and validate that
+	// volume directory is not directly under root.
+	// Extract the base path and the volume unique path.
+	var vErr error
+	config.parentDir, config.volumeDir, vErr = hostpath.NewBuilder().WithPath(pOpts.path).
+		WithCheckf(hostpath.IsNonRoot(), "volume directory {%v} should not be under root directory", pOpts.path).
+		ExtractSubPath()
+	if vErr != nil {
+		return vErr
+	}
+
+	// Pass on the taints, to create tolerations.
+	config.taints = pOpts.selectedNodeTaints
+
+	var lErr error
+	config.pOpts.softLimitGrace, lErr = convertToK(config.pOpts.softLimitGrace, config.pOpts.pvcStorage)
+	if lErr != nil {
+		return lErr
+	}
+	config.pOpts.hardLimitGrace, lErr = convertToK(config.pOpts.hardLimitGrace, config.pOpts.pvcStorage)
+	if lErr != nil {
+		return lErr
+	}
+
+	if err := pOpts.validateLimits(); err != nil {
+		return err
+	}
+
+	// fs stores the file system of mount
+	fs := "FS=`stat -f -c %T /data` ; "
+	// check if fs is xfs or ext4 (output of stat is ext2/ext3)
+	// PID is the last project Id in the directory
+	// xfs_quota project(xfs) or chattr +P (ext4) initializes project with new project id
+	// xfs_quota limit(xfs) or repquota (ext4) sets the quota according to limits defined
+	checkQuota := "set -x;" +
+		"if [[ \"$FS\" == \"xfs\" ]]; then " +
+		"  PID=`xfs_quota -x -c 'report -h' /data | tail -2 | awk 'NR==1{print substr ($1,2)}+0'` ;" +
+		"  PID=`expr $PID + 1` ;" +
+		"  xfs_quota -x -c 'project -s -p " + filepath.Join("/data/", config.volumeDir) + " '$PID /data;" +
+		"  xfs_quota -x -c 'limit -p bsoft=" + config.pOpts.softLimitGrace + " bhard=" + config.pOpts.hardLimitGrace + " '$PID /data ;" +
+		"elif [[ \"$FS\" == \"ext2/ext3\" ]]; then" +
```

Comment:
ext4？

Reply:
> // check if fs is xfs or ext4 (output of stat is ext2/ext3)

for ext4, the output of stat is "ext2/ext3".